### PR TITLE
use `run_shell_cmd` in custom easyblock for ABAQUS

### PR DESCRIPTION
**Pull request by boegel**

### CODE REVIEW

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _ABAQUS-2022-hotfix-2214.eb_ (partial log available at https://gist.github.com/boegel/5d13c7c0cabb3197beb35bccfa903ade)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
node3560.doduo.os - Linux RHEL 8.8, x86_64, AMD EPYC 7552 48-Core Processor, Python 3.6.8
See https://gist.github.com/boegel/515075a30ef324af70708ade0e09b5f2 for a full test report.

**Comment by boegel** [comment]
Hitting an interesting problem when testing with `ABAQUS-2022-hotfix-2214.eb`.

At some point `run_shell_cmd` checks for question pattern matches in following end of stdout:
```
== 2024-08-20 21:10:00,492 run.py:270 DEBUG Found match for question pattern '\s*(?P<nr>[-0-9]+)[\s\n]+\[\*\][\s\n]+Isight[\s\w]*\n((?!Enter[\s\n]+selection[\s\n]+\(default:[\s\n]+Next\):\s*)(?!___)[\S[\s\n]+]*\n)*Enter[\s\n]+selection[\s\n]+\(default:[\s\n]+Next\):\s*$[\s\n]*$' at end of stdout: lished Products CAA API
  7 [*] Isight

Enter selection (default: Next):
 -1 Select All
  0 Select None
  1 [ ] NETVIBES Exalead CloudView
  2 [ ] EXALEAD Search for Documentation
  3 [ ] Extended Product Documentation
  4 [ ] FLEXnet License Server
  5 [*] SIMULIA Established Products
  6 [*] SIMULIA Established Products CAA API
  7 [ ] Isight

Enter selection (default: Next):
 -1 Select All
  0 Select None
  1 [ ] NETVIBES Exalead CloudView
  2 [ ] EXALEAD Search for Documentation
  3 [ ] Extended Product Documentation
  4 [ ] FLEXnet License Server
  5 [*] SIMULIA Established Products
  6 [*] SIMULIA Established Products CAA API
  7 [*] Isight

Enter selection (default: Next):
 -1 Select All
  0 Select None
  1 [ ] NETVIBES Exalead CloudView
  2 [ ] EXALEAD Search for Documentation
  3 [ ] Extended Product Documentation
  4 [ ] FLEXnet License Server
  5 [*] SIMULIA Established Products
  6 [*] SIMULIA Established Products CAA API
  7 [ ] Isight

Enter selection (default: Next):
```

The problem here is that "`\[\*\] Isight`" matches, even though it's not at the end of the output.
It should be matched once, to disable the installation of `Isight`.
But now we're going into an infinite enable/disable loop, because `run_shell_cmd` considers the *full* output, not just the last part of the output...

This probably indicates that another bug fix is required after https://github.com/easybuilders/easybuild-framework/pull/4617, to only consider the new part of the output after a question has been answered?

**Comment by boegel** [comment]
Problem with the tests has nothing to do with the changes to the ABAQUS easyblock, it's fallout from a change in framework that broke the easyblocks test suite (I'll get that fixed in a separate PR):
* https://github.com/easybuilders/easybuild-framework/pull/4585

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ABAQUS-2022-hotfix-2214.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3515.doduo.os - Linux RHEL 8.8, x86_64, AMD EPYC 7552 48-Core Processor, Python 3.6.8
See https://gist.github.com/boegel/05fca2a4aafc3660f0ad6dd2878a33bc for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _ABAQUS-2023-hotfix-2324.eb_ (partial log available at https://gist.github.com/Micket/e331163df66c00c9d5cb76c60bd9c33f)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
alvis-skylake-build - Linux Rocky Linux 8.9, x86_64, Intel Xeon Processor (Skylake, IBRS, no TSX), Python 3.6.8
See https://gist.github.com/Micket/f15b99bfca3191a72ff3bdf981ea8e43 for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ABAQUS-2022-hotfix-2223.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
alvis-skylake-build - Linux Rocky Linux 8.9, x86_64, Intel Xeon Processor (Skylake, IBRS, no TSX), Python 3.6.8
See https://gist.github.com/Micket/cfd93697b9f9af61cecf227c8a8297be for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ABAQUS-2021.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0105u03a - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/branfosj/86fac84a70f2d1e356b660e403e9e471 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ABAQUS-2024-hotfix-2405.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0105u03b - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/branfosj/61d9dbdfbb3d4090f27573ee2cf4b226 for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _ABAQUS-2023-hotfix-2324.eb_ (partial log available at https://gist.github.com/Micket/f9d6796f8b8b7985d429b918ebd80690)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
alvis-skylake-build - Linux Rocky Linux 8.9, x86_64, Intel Xeon Processor (Skylake, IBRS, no TSX), Python 3.6.8
See https://gist.github.com/Micket/c8ac16672d415e8d38dc0219e1503d03 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ABAQUS-2023-hotfix-2314.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0105u03a - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/branfosj/98d04f891fa477887b8cf7c910f7b5c3 for a full test report.

File: easybuild/easyblocks/a/abaqus.py [hunk 1]
```diff
@@ -255,7 +253,7 @@ def install_step(self):
                         raise EasyBuildError("Failed to find expected subdir for hotfix: %s", subdirs)
 
                     cwd = change_dir(os.path.join(cwd, subdir, '1'))
-                    run_cmd_qa('./StartTUI.sh', {}, std_qa=std_qa, log_all=True, simple=True, maxhits=100)
+                    run_shell_cmd('./StartTUI.sh', qa_patterns=qa, qa_wait_patterns=no_qa, qa_timeout=100)
```
Comment (branfosj) [inline easybuild/easyblocks/a/abaqus.py:256]:
```suggestion
                    run_shell_cmd('./StartTUI.sh', qa_patterns=qa, qa_wait_patterns=no_qa, qa_timeout=200)
```

We may need to make this longer.

File: easybuild/easyblocks/a/abaqus.py [hunk 2]
```diff
@@ -294,18 +293,19 @@ def install_step(self):
...
-                    run_cmd_qa('./StartTUI.sh', {}, no_qa=no_qa, std_qa=std_qa, log_all=True, simple=True, maxhits=100)
+                    qa = [
+                        (r"Enter selection \(default: Next\):", ''),
+                        (r"Choose the .*installation directory.*\n.*\n\n.*:", os.path.join(self.installdir, 'cae')),
+                        (r"Enter selection \(default: Install\):", ''),
+                        (r"The Abaqus commands directory.*:\n.*\n+Actions:\n.*\n_+\n\nPlease.*:", '1'),
+                        (r"Enter selection \(default: Close\):", ''),
+                    ]
+
+                    run_shell_cmd('./StartTUI.sh', qa_patterns=qa, qa_wait_patterns=no_qa, qa_timeout=100)
```
Comment (branfosj) [inline easybuild/easyblocks/a/abaqus.py:308]:
```suggestion
                    run_shell_cmd('./StartTUI.sh', qa_patterns=qa, qa_wait_patterns=no_qa, qa_timeout=200)
```

**Comment by branfosj** [comment]
`qa_timeout=100` is the default value. We should increase any that need increasing and remove the others.

**Review by Micket — APPROVED** [review]
LGTM (I'm fine with 1000 everywhere. Probably overkill for many of these, but who knows, still isn't to terrible to wait out 1000 if there truly is some situation where it is broken)

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ABAQUS-2022-hotfix-2214.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3515.doduo.os - Linux RHEL 8.8, x86_64, AMD EPYC 7552 48-Core Processor, Python 3.6.8
See https://gist.github.com/boegel/3d7362b49834df6f2a15c0cddbd27bb3 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ABAQUS-2021.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0105u03a - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/branfosj/fda52432711b771d7164d3a2b418cb66 for a full test report.

**Comment by branfosj** [comment]
Going in, thanks @boegel!

**Comment by branfosj** [comment]
Test report by @branfosj
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ABAQUS-2024-hotfix-2405.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0105u03a - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/branfosj/deda4dc2ff7a74383102bc64a264b00c for a full test report.

**Pull request merged**